api/rbody/http_response.go

```diff
@@ -0,0 +1,213 @@
+package rbody
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+
+	"github.com/raintank/metrictank/util"
+	"github.com/tinylib/msgp/msgp"
+)
+
+type FastJSON interface {
+	MarshalJSONFast([]byte) ([]byte, error)
+}
+
+var ErrMetricNotFound = errors.New("metric not found")
+
+func WriteResponse(w http.ResponseWriter, resp Response) {
+	defer resp.Close()
+	body, err := resp.Body()
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(err.Error()))
+		return
+	}
+	for k, v := range resp.Headers() {
+		w.Header().Set(k, v)
+	}
+	w.WriteHeader(resp.Code())
+	w.Write(body)
+	return
+}
+
+type Response interface {
+	Code() int
+	Body() ([]byte, error)
+	Headers() map[string]string
+	Close()
+}
+
+type JsonResponse struct {
+	code  int
+	body  interface{}
+	jsonp string
+}
+
+func NewJsonResponse(code int, body interface{}, jsonp string) *JsonResponse {
```

Comment:
For consistency all "Response" types should be the same.  FastJsonResponse, MsgpResponse and MsgpArrayResponse all need to use pointers as the Body() method mutates the struct.  So all implementations of the Response interface 

https://play.golang.org/p/U5Z3t3YsPl

And also
https://github.com/golang/go/wiki/CodeReviewComments#receiver-type